apps/common/controllers/base-address.js

```diff
@@ -0,0 +1,50 @@
+'use strict';
+
+const BaseController = require('./base');
+const _ = require('lodash');
+
+module.exports = class BaseAddressController extends BaseController {
+
+  locals(req, res) {
+    const locals = super.locals(req, res);
+    let addresses = req.sessionModel.get(this.options.addressKey);
+    const hasAddresses = !!_.size(addresses);
+    const hasCategories = this.hasCategories(hasAddresses, addresses);
+    if (this.options.addressKey === 'locationAddresses') {
+      addresses = _.filter(addresses, value => value.categories !== undefined);
+    }
+    const items = this.mapAddresses(addresses, req);
+
+    return Object.assign({}, locals, {
+      items,
+      hasAddresses,
+      hasCategories
+    });
+  }
+
+  translateCategories(req, values) {
+    return _.castArray(values).map(value => {
+      const key = `fields.location-address-category.options.${value}.label`;
+      const result = req.translate(key);
+      return result === key ? value : result;
+    }).join('\n');
+  }
+
+  mapAddresses(addresses, req) {
+    return _.map(addresses, (value, key) => ({
+      id: key,
+      address: value.address,
+      categories: this.translateCategories(req, value.categories)
+    }));
+  }
+
+  hasCategories(hasAddresses, addresses) {
+    if (hasAddresses) {
+      const firstAddressItem = addresses[Object.keys(addresses)[0]];
+      return _.has(firstAddressItem, 'categories');
+    }
+    else {
+      return false;
+    }
+  }
+};
```

Comment:
to understand this a bit further, addresses is made up like so
addresses>> { '2':
   { address: 'Paris',
     postcode: '',
     categories: [ 'small-bore-rifles', 'muzzle-loading-pistols' ] },
  '3':
   { address: 'Jordan',
     postcode: '',
     categories:
      [ 'full-bore-rifles',
        'small-bore-rifles',
        'muzzle-loading-pistols' ] } }
hasCategories just checks that any child object has a property called categories.  I've attempted to make it more readable and not use the _.sample function

Let me know if it can be improved